ciao-controller/internal/datastore/datastore_test.go

```diff
@@ -277,6 +285,36 @@ func BenchmarkAllocateTenantIP(b *testing.B) {
 	}
 }
 
+func BenchmarkAllocate1000TenantIP(b *testing.B) {
```

Comment:
Are you concerned about performance of the IP allocator?  In some ways this isn't a good test as you are starting with a clean tenant with no IPs allocated and so is the best case for the allocator.

Reply:
If you are concerned about performance it may be possible to speed things up a little.  What operations do you perform on the  ds.tenants[tenantID].network[subnet] maps?  Does this map just exist to keep track of allocations?  Perhaps you could use a sorted slice to keep track of which ips are and are not allocated, with the allocated ips coming first, and the free ips second.  Allocation then becomes O(N) where N is the number of subnets.  De-allocation may be slower but in the worst case its a 1KB memmove.

Reply:
Or you could keep the existing algorithm but replace the make(map[uint32]bool) with a bitmask backed by big.Int.  This would use much less memory and your lookups, sets and deletes would be faster as there would be no allocations and you should have better cache performance.

Reply:
Or each subnet could have two maps or big.Ints.  One that records allocated ips and one which records free ips.  To allocate ips from a subnet you just need to iterate through the free list and move them to the allocated list.  If nothing is free you skip the subnet.  No need to iterate through it.

Reply:
I am concerned with performance in that I don't want to regress (which this algorithm does due to the translations from addresses to bytes that didn't happen as much in the old algorithm). I can play with some of these ideas and see if they help significantly.

Reply:
@markdryan I have come to the conclusion that working on optimizing this algorithm isn't the best use of time right now, as the larger issue during launch are the multiple sql accesses that still exist outside of the network allocation. Removing the sql table for tenant_network and parallelizing the launch seem to be a good improvement for now, and we can optimize further in the future.